src/sentry/models/projectsdk.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import annotations
+
+import logging
+from collections.abc import Sequence
+from enum import Enum
+
+from django.db import models
+from packaging.version import InvalidVersion, Version
+from packaging.version import parse as parse_version
+
+from sentry.backup.scopes import RelocationScope
+from sentry.db.models import BoundedIntegerField, FlexibleForeignKey, region_silo_model, sane_repr
+from sentry.db.models.base import DefaultFieldsModel
+from sentry.locks import locks
+from sentry.models.project import Project
+from sentry.sdk_updates import get_sdk_index
+from sentry.utils import metrics
+from sentry.utils.cache import cache
+from sentry.utils.hashlib import md5_text
+
+logger = logging.getLogger(__name__)
+
+
+class EventType(Enum):
+    PROFILE = 0
+    PROFILE_CHUNK = 1
+
+    @classmethod
+    def as_choices(cls) -> Sequence[tuple[int, str]]:
+        return (
+            (cls.PROFILE.value, "profile"),
+            (cls.PROFILE_CHUNK.value, "profile_chunk"),
+        )
+
+
+@region_silo_model
+class ProjectSDK(DefaultFieldsModel):
+    __relocation_scope__ = RelocationScope.Organization
+
+    date_updated = models.DateTimeField(auto_now=True)
+
+    project = FlexibleForeignKey("sentry.Project", on_delete=models.CASCADE)
+    event_type = BoundedIntegerField(choices=EventType.as_choices())
+    sdk_name = models.CharField()
+    sdk_version = models.CharField()
+
+    class Meta:
+        unique_together = (("project", "event_type", "sdk_name"),)
+
+    __repr__ = sane_repr("project", "event_type", "sdk_name", "sdk_version")
+
+    @classmethod
+    def get_lock_key(cls, project: Project, event_type: EventType, sdk_name: str):
+        return f"lprojectsdk:{project.id}:{event_type.value}:{md5_text(sdk_name).hexdigest()}"
+
+    @classmethod
+    def get_cache_key(cls, project: Project, event_type: EventType, sdk_name: str):
+        return f"projectsdk:{project.id}:{event_type.value}:{md5_text(sdk_name).hexdigest()}"
+
+    @classmethod
+    def update_with_newest_version_or_create(
+        cls,
+        project: Project,
+        event_type: EventType,
+        sdk_name: str,
+        sdk_version: str,
+    ):
+        try:
+            new_version = parse_version(sdk_version)
+        except InvalidVersion:
+            # non-semver sdk version, ignore and move on
+            return
+
+        normalized_sdk_name = normalize_sdk_name(sdk_name)
+        if normalized_sdk_name is None:
+            logger.info("Unknown sdk name: %s", sdk_name)
+            return
+
+        lock_key = cls.get_lock_key(project, event_type, normalized_sdk_name)
+        lock = locks.get(lock_key, duration=10, name="projectsdk")
+
+        # This can raise `sentry.utils.locking.UnableToAcquireLock`
+        # that needs to be handled by the caller and handled
+        # appropriately with retries if needed.
```

Comment:
You could also just catch it in here, but I guess maybe that behaviour is a bit unclear and less flexible

Reply:
Yeah, I thought about catching it here but that means this method can now silently fail and not update anything which seems like a foot gun.